v2/spiffeid/id.go

```diff
@@ -1,78 +1,162 @@
 }
 
 // MustJoin returns the string representation of an ID inside the given trust
 // domain (e.g. example.org) with the given path segments. The function panics
 // if the trust domain is not valid (see TrustDomainFromString).
 func MustJoin(trustDomain string, segments ...string) string {
-	panic("not implemented")
+	idstr, err := Join(trustDomain, segments...)
+	if err != nil {
+		panic(err)
+	}
+	return idstr
 }
 
 // FromString parses a SPIFFE ID from a string.
 func FromString(s string) (ID, error) {
-	panic("not implemented")
+	uri, err := url.Parse(s)
+	if err != nil {
+		return ID{}, fmt.Errorf("invalid SPIFFE ID: %v", err)
+	}
+
+	return FromURI(uri)
 }
 
 // FromURI parses a SPIFFE ID from a URI.
-func FromURI(u *url.URL) (ID, error) {
-	panic("not implemented")
+func FromURI(uri *url.URL) (ID, error) {
+	if uri == nil || *uri == (url.URL{}) {
+		return ID{}, errors.New("invalid SPIFFE ID: SPIFFE ID is empty")
+	}
+
+	// General validation
+	switch {
+	case strings.ToLower(uri.Scheme) != "spiffe":
+		return ID{}, errors.New("invalid SPIFFE ID: invalid scheme")
+	case uri.User != nil:
+		return ID{}, errors.New("invalid SPIFFE ID: user info is not allowed")
+	case uri.Host == "":
+		return ID{}, errors.New("invalid SPIFFE ID: trust domain is empty")
+	case uri.Port() != "":
+		return ID{}, errors.New("invalid SPIFFE ID: port is not allowed")
+	case uri.Fragment != "":
+		return ID{}, errors.New("invalid SPIFFE ID: fragment is not allowed")
+	case uri.RawQuery != "":
+		return ID{}, errors.New("invalid SPIFFE ID: query is not allowed")
+	}
+
+	uri = normalizeURI(uri)
+
+	return ID{
+		td:   TrustDomain{name: uri.Host},
+		path: uri.EscapedPath(),
+	}, nil
 }
 
 // TrustDomain returns the trust domain of the SPIFFE ID.
 func (id ID) TrustDomain() TrustDomain {
-	panic("not implemented")
+	return id.td
 }
 
 // MemberOf returns true if the SPIFFE ID is a member of the given trust domain.
 func (id ID) MemberOf(td TrustDomain) bool {
-	panic("not implemented")
+	return id.td.name == td.name
 }
 
 // Path returns the path of the SPIFFE ID inside the trust domain.
 func (id ID) Path() string {
-	panic("not implemented")
+	return id.path
 }
 
 // String returns the string representation of the SPIFFE ID, e.g.,
 // "spiffe://example.org/foo/bar".
 func (id ID) String() string {
-	panic("not implemented")
+	if id.Empty() {
+		return ""
+	}
+
+	return id.URL().String()
 }
 
 // URL returns a URL for SPIFFE ID.
 func (id ID) URL() *url.URL {
-	panic("not implemented")
+	if id.Empty() {
+		return &url.URL{}
+	}
+
+	return &url.URL{
+		Scheme: "spiffe",
+		Host:   id.td.name,
+		Path:   id.path,
+	}
 }
 
 // Empty returns true if the SPIFFE ID is empty.
 func (id ID) Empty() bool {
-	panic("not implemented")
+	return id.td.name == ""
+}
+
+func normalizeURI(uri *url.URL) *url.URL {
+	c := *uri
+	c.Scheme = strings.ToLower(c.Scheme)
+	// SPIFFE ID's can't contain ports so don't bother handling that here.
+	c.Host = strings.ToLower(c.Hostname())
```

Comment:
The use of `Hostname()` could mask a bug in the validation code (like I mentioned with the `:` delimiter followed by no port). I think we should stick with `c.Host` so that it's obvious something was wrong from the unit-tests.